crates/pathfinder/examples/re_execute.rs

```diff
@@ -0,0 +1,556 @@
+
+    let chain = match genesis_hash {
+        MAINNET_GENESIS_HASH => ChainId::MAINNET,
+        TESTNET_GENESIS_HASH => ChainId::TESTNET,
+        TESTNET2_GENESIS_HASH => ChainId::TESTNET2,
+        INTEGRATION_GENESIS_HASH => ChainId::INTEGRATION,
+        _ => anyhow::bail!("Unknown chain"),
+    };
+
+    Ok(chain)
+}
+
+#[derive(Debug)]
+struct Work {
+    block_number: BlockNumber,
+    block_timestamp: BlockTimestamp,
+    sequencer_address: SequencerAddress,
+    state_at_block: Option<BlockNumber>,
+    gas_price: U256,
+    transactions: Vec<Transaction>,
+    receipts: Vec<Receipt>,
+}
+
+fn execute(storage: Storage, chain_id: ChainId, rx: crossbeam_channel::Receiver<Work>) {
+    while let Ok(work) = rx.recv() {
+        let start_time = std::time::Instant::now();
+        let num_transactions = work.transactions.len();
+
+        let connection = storage.connection().unwrap();
+
+        let mut execution_state = ExecutionState {
+            connection,
+            chain_id,
+            block_number: work.block_number,
+            block_timestamp: work.block_timestamp,
+            sequencer_address: work.sequencer_address,
+            state_at_block: work.state_at_block,
+            gas_price: work.gas_price,
+            pending_update: None,
+        };
+
+        let db_tx = execution_state
+            .connection
+            .transaction()
+            .expect("Create transaction");
+
+        let transactions = work
+            .transactions
+            .into_iter()
+            .map(|tx| map_gateway_transaction(tx, &db_tx))
+            .collect::<Result<Vec<_>, _>>();
+
+        drop(db_tx);
+
+        let transactions = match transactions {
+            Ok(transactions) => transactions,
+            Err(error) => {
+                tracing::error!(block_number=%work.block_number, %error, "Transaction conversion failed");
+                continue;
+            }
+        };
+
+        match pathfinder_executor::estimate(execution_state, transactions) {
+            Ok(fee_estimates) => {
+                for (estimate, receipt) in fee_estimates.iter().zip(work.receipts.iter()) {
+                    if let Some(actual_fee) = receipt.actual_fee {
+                        let actual_fee = u128::from_be_bytes(
+                            actual_fee.0.to_be_bytes()[16..].try_into().unwrap(),
+                        );
+
+                        // L1 handler transactions have a fee of zero in the receipt.
+                        if actual_fee == 0 {
+                            continue;
+                        }
+
+                        let gas_price = work.gas_price.as_u128();
+                        let actual_gas_consumed = actual_fee / gas_price.max(1);
+
+                        let estimated_gas_consumed = estimate.gas_consumed.as_u128();
+
+                        let diff = actual_gas_consumed.abs_diff(estimated_gas_consumed);
+
+                        if diff > (actual_gas_consumed * 2 / 10) {
```

Comment:
This 20% is pretty arbitrary. I had a look at how [starkli](https://github.com/xJonathanLEI/starkli) (and presumably other tools) submits transactions:

- First the transaction is constructed and sent to `starknet_estimateFee` so that the submitter has a guesstimate of how much submitting the transaction costs.
- The value returned is not guaranteed to be accurate: since gas price fluctuates and the final costs also depend on the number of storage updates the returned value is not used directly. Instead, starkli [multiplies the value by 1.5](https://github.com/xJonathanLEI/starkli/blob/4c827d96f7420e1abf7f95f4f5723cd1ae99d733/src/subcommands/invoke.rs#L120) and uses that as the `max_fee` for the transaction.

My thinking was that if the results are within 20% of the actual value charged by the sequencer then the estimate is good enough so that the transaction would be accepted after multiplying by 1.5.

Reply:
Understood thanks :bow:

Reply:
There's another source of non-exact estimates: even with blockifier things like syscall costs are updated regularly, so just like with cairo-lang each estimate we give is only "accurate" (considering the things above) for the version of transactions submitted for the most recent Starknet version (that uses the same costs / weights).